### PR TITLE
[Perf]: Fused RMSNorm+3D Rope kernel in Sensenova-U1

### PR DESCRIPTION
**Pull request by sphinxkkkbc**
PLEASE FILL IN THE PR DESCRIPTION HERE.

## Purpose

Implements part of #3514: fuse RMSNorm + 3D RoPE in `SenseNovaU1Attention._project_and_rope`.

The previous implementation follows the PyTorch reference path:

1. split Q/K into `t` and `hw`
2. run RMSNorm separately
3. transpose to attention layout
4. split `hw` into `h` and `w`
5. apply 3D RoPE
6. concatenate Q/K states

This creates multiple small PyTorch ops and visible dtype conversion/copy overhead, especially from `aten::to`, `aten::_to_copy`, and `aten::copy_`. This PR adds a Triton fused kernel for the Q/K RMSNorm + 3D RoPE path while keeping the PyTorch fallback.

**Note:** A single run includes many steps and layers, and profiler traces vary across them. The screenshots below show representative examples from the common path.

**vLLM Version:** 0.23.0  
**vLLM-Omni Commit:** https://github.com/vllm-project/vllm-omni/pull/4669/commits/bb82c7a0cc375b3ae9be0eec14cdeca36bfa9e89

## Profile of Native Implementation

The torch profiler shows visible `aten::to` / `aten::_to_copy` / `aten::copy_` overhead across attention layers. The screenshots below include both the denoising path and the `ar_step` path. The same copy overhead exists in both paths, but its relative contribution is smaller in `ar_step`; it is more visible in the denoising path.

#### Denoising step

image

#### AR step

image

## Fuse Logic

The fused kernel computes the same math as the native PyTorch path for Q/K:

```python
q_t, q_hw = q.chunk(2, dim=-1)
k_t, k_hw = k.chunk(2, dim=-1)

q_t = q_norm(q_t)
k_t = k_norm(k_t)
q_hw = q_norm_hw(q_hw)
k_hw = k_norm_hw(k_hw)

q_h, q_w = q_hw.chunk(2, dim=-1)
k_h, k_w = k_hw.chunk(2, dim=-1)

q_t, k_t = apply_rotary_pos_emb(q_t, k_t, cos_t, sin_t)
q_h, k_h = apply_rotary_pos_emb(q_h, k_h, cos_h, sin_h)
q_w, k_w = apply_rotary_pos_emb(q_w, k_w, cos_w, sin_w)

query_states = torch.cat([q_t, q_h, q_w], dim=-1)
key_states = torch.cat([k_t, k_h, k_w], dim=-1)
```

The Triton kernel fuses these operations into one pass per `(batch, token, head)` for Q/K:

- computes RMS scales for the `t` half and `hw` half in fp32
- applies the corresponding Q/K RMSNorm weights
- applies RoPE for `t`, `h`, and `w`
- writes directly to `[B, H, S, D]` attention layout

The implementation keeps a PyTorch fallback path when the fused kernel is unavailable.

I also explored a `without_pair` formulation that computes each RoPE pair directly and writes two outputs per lane. Local kernel-level smoke tests showed only small gains for short sequence lengths and regressions for longer sequence lengths, so the current default keeps the pair-based formulation.

#### Denoising step

Image

#### AR step

Image

## Test Plan

Correctness test against the native PyTorch reference:

```bash
python3 -m pytest tests/diffusion/models/sensenova_u1/test_triton_kernels.py
```

Performance benchmark:

```bash
python3 benchmarks/diffusion/diffusion_benchmark_serving.py \
  --base-url http://localhost:8091 \
  --endpoint /v1/chat/completions \
  --model SenseNova/SenseNova-U1-8B-MoT \
  --task t2i \
  --dataset random \
  --num-prompts 10 \
  --warmup-requests 2 \
  --warmup-num-inference-steps 5 \
  --max-concurrency 1 \
  --width 2048 \
  --height 2048 \
  --num-inference-steps 50 \
  --seed 42 \
  --output-file sensenova_u1_bench.json
```

## Test Result

After using the fused kernel, `Qwen3RotaryEmbedding` becomes the next visible bottleneck and can be optimized separately.

| Metric | fused_kernel | native | Change |
|---|---:|---:|---:|
| Latency Mean | 49.2403 s | 55.3656 s | -11.1% |
| Latency Median | 49.2951 s | 55.5653 s | -11.3% |
| Latency P95 | 49.6381 s | 55.7862 s | -11.0% |
| Latency P99 | 49.6556 s | 55.7871 s | -11.0% |
| Stage 0 Gen Mean | 47.1002 s | 53.2120 s | -11.5% |
| Peak Memory Max | 36490 MB | 36522 MB | -32 MB |

### Numerical Differences

The fused kernel matches the PyTorch reference in fp32. In bf16, small numerical differences are expected because the fused path changes the order of bf16 casts, multiply-adds, and stores compared with the PyTorch op sequence.

The table below reports absolute error between fused output and PyTorch native output for bf16:

| Seq_len | max | p99 | mean |
|---:|---:|---:|---:|
| 1 | 0.062500 | 0.015625 | 0.000829 |
| 2 | 0.062500 | 0.015625 | 0.000817 |
| 4 | 0.062500 | 0.015625 | 0.000875 |
| 8 | 0.062500 | 0.015625 | 0.000793 |
| 9 | 0.062500 | 0.015625 | 0.000744 |
| 16 | 0.125000 | 0.015625 | 0.000853 |
| 32 | 0.062500 | 0.015625 | 0.000825 |
| 64 | 0.062500 | 0.015625 | 0.000790 |
| 128 | 0.125000 | 0.015625 | 0.000805 |
| 256 | 0.125000 | 0.015625 | 0.000869 |
| 260 | 0.125000 | 0.015625 | 0.000778 |
| 271 | 0.125000 | 0.015625 | 0.000740 |
| 512 | 0.125000 | 0.015625 | 0.000797 |
| 1024 | 0.125000 | 0.015625 | 0.000761 |
| 2048 | 0.125000 | 0.015625 | 0.000721 |
| 4096 | 0.125000 | 0.015625 | 0.000754 |

The p99 and mean errors stay stable across sequence lengths. The max error is caused by sparse bf16 outliers.

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py [hunk 1]
```diff
@@ -373,6 +376,34 @@ def _project_and_rope(self, hidden_states, indexes, qkv_proj, q_norm, k_norm, q_
...
+                sin_t,
+                cos_h,
+                sin_h,
+                cos_w,
+                sin_w,
+                self.config.rms_norm_eps,
+            )
+        else:
+            logger.debug("Triton fused SenseNova qk norm rope unavailable; using PyTorch fallback")
+
```
[the remark this answers is not in the document]
Comment (sphinxkkkbc) [reply vllm_omni/diffusion/models/sensenova_u1/sensenova_u1_transformer.py:406]: the benchmark was collected from the A/B test branch. During a later refactor that merged the two branches into one function, the fused path initially missed an early return and could fall through to the native fallback. This has been fixed in the latest commit.

**Review by hsliuustc0106 — COMMENTED** [review]
Fused RMSNorm+3D RoPE kernel looks solid. Correctness tests cover fp32 and bf16 numerical differences, performance benchmarks show 11% latency reduction. Gate checks pass.

**Review by princepride — APPROVED** [review]
